zenoh-flow/src/traits.rs

```diff
@@ -22,49 +22,194 @@ use std::any::Any;
+///
+/// Example:
+/// ```no_run
+///
+/// use zenoh_flow::{Deserializable, ZFResult, ZFError};
+/// use zenoh_flow::zenoh_flow_derive::ZFData;
+///
+/// #[derive(Debug, Clone, ZFData)]
+/// pub struct MyString(pub String);
+///
+/// impl Deserializable for MyString {
+///     fn try_deserialize(bytes: &[u8]) -> ZFResult<MyString>
+///     where
+///         Self: Sized,
+///     {
+///         Ok(MyString(
+///             String::from_utf8(bytes.to_vec()).map_err(|_| ZFError::DeseralizationError)?,
+///         ))
+///     }
+/// }
+/// ```
 pub trait Deserializable {
+    /// Tries to deserialize from a slice of `u8`.
+    ///
+    /// # Errors
+    /// If it fails to deserialize an error variant will be returned.
     fn try_deserialize(bytes: &[u8]) -> ZFResult<Self>
     where
         Self: Sized;
 }
 
+/// This trait abstracts the user's state type inside Zenoh Flow.
+///
+/// User types should implement this trait otherwise Zenoh Flow will
+/// not be able to handle the state.
+///
+/// It can be easily derived.
+///
+/// Example:
+///
+/// ```no_run
+/// use zenoh_flow::zenoh_flow_derive::ZFState;
+/// #[derive(Debug, Clone, ZFState)]
+/// pub struct MyState;
+/// ```
+///
+
 pub trait ZFState: Debug + Send + Sync {
+    /// Donwcast as reference to [`Any`](`Any`)
     fn as_any(&self) -> &dyn Any;
+
+    /// Donwcast as mutable reference to [`Any`](`Any`)
     fn as_mut_any(&mut self) -> &mut dyn Any;
 }
 
+/// The `Node` trait represents a generic node in the data flow graph.
+/// It contains functions that are common between Operator, Sink and Source.
+/// It has to be implemented for each node within a graph.
 pub trait Node {
+    /// This method is used to initialize the state of the node.
+    /// It is called by the Zenoh Flow runtime when initializing the data flow
+    /// graph.
+    /// An example of node state is files that should be opened, connection
+    /// to devices or internal configuration.
+    ///
+    /// # Errors
+    /// If it fails to initialize an error variant will be returned.
     fn initialize(&self, configuration: &Option<Configuration>) -> ZFResult<State>;
 
+    /// This method is used to finalize the state of the node.
+    /// It is called by the Zenoh Flow runtime when tearing down the data
+    /// flow graph.
+    ///
+    /// An example of node state to finalize is files to be closed,
+    /// clean-up of libraries, or devices.
+    ///
+    /// # Errors
+    /// If it fails to finalize an error variant will be returned.
     fn finalize(&self, state: &mut State) -> ZFResult<()>;
 }
 
+/// The `Operator` trait represent an Operator inside Zenoh Flow.
 pub trait Operator: Node + Send + Sync {
+    /// This method is called when data is received on one or more inputs.
+    /// The result of this method is use as discriminant to trigger the
+    /// operator's run function.
+    /// The operator can access to its context and its state during execution.
```

Comment:
An operator can access its context and state during the execution of this function.